src/Elliptic/Systems/Poisson/Observe.hpp

```diff
@@ -0,0 +1,124 @@
+#include "IO/Observer/Actions.hpp"
+#include "IO/Observer/ArrayComponentId.hpp"
+#include "IO/Observer/ObservationId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/ReductionActions.hpp"
+#include "IO/Observer/VolumeActions.hpp"
+#include "NumericalAlgorithms/LinearOperators/DefiniteIntegral.hpp"
+#include "NumericalAlgorithms/LinearSolver/Tags.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/MakeString.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Poisson {
+namespace Actions {
+
+/*!
+ * \brief Temporary action to observe volume and reduction data for the Poisson
+ * system
+ */
+struct Observe {
+  template <typename... DbTags, typename... InboxTags, typename Metavariables,
+            size_t Dim, typename ActionList, typename ParallelComponent>
+  static auto apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ElementIndex<Dim>& array_index,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& iteration_id = get<LinearSolver::Tags::IterationId>(box);
+    const auto& mesh = get<Tags::Mesh<Dim>>(box);
+    const std::string element_name = MakeString{} << ElementId<Dim>(array_index)
+                                                  << '/';
+
+    // Retrieve the current numeric solution
+    const auto& field = get<Poisson::Field>(box);
+
+    // Compute the analytic solution
+    const auto& inertial_coordinates =
+        db::get<::Tags::Coordinates<Dim, Frame::Inertial>>(box);
+    const auto field_analytic = get<Poisson::Field>(
+        Parallel::get<typename Metavariables::analytic_solution_tag>(cache)
+            .field_variables(inertial_coordinates));
+
+    // Compute error between numeric and analytic solutions
+    DataVector field_error = get(field) - get(field_analytic);
+
+    // Retrieve linear residual
+    // We have to reduce it here again, even though it should be the same on all
+    // elements, because it was computed in the linear solver residual monitor
+    // singleton. Ideally we would contribute it to the observation from the
+    // singleton when that becomes possible.
+    using fields_tag = typename Metavariables::system::fields_tag;
+    using residual_magnitude_tag = db::add_tag_prefix<
+        LinearSolver::Tags::Magnitude,
+        db::add_tag_prefix<LinearSolver::Tags::Residual, fields_tag>>;
+    const double linear_residual = get<residual_magnitude_tag>(box);
+
+    // Compute l2 error square over local element
+    const double local_l2_error_square =
+        definite_integral(square(field_error), mesh);
```

Comment:
I have noticed that we compute L2 norms differently in other Observers, but this method of computing the L2 norm does not seem unreasonable to me, so I don't have an immediate objection. However, I think we could gain from carefully revisiting how L2 norms are computed numerically, ideally with a source from the literature if someone knows offhand where it is described, so that we can all be on the same page and provide future people with a reference if they have a question about this is well.

Reply:
I think so too and actually opened #1189 for this.